### PR TITLE
AES128GCM implementation

### PR DESCRIPTION
**Pull request by eoger**
@linacambridge mind having a look? :)

### CODE REVIEW

File: src/aes128gcm.rs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+            pad_length,
+            salt: Some(salt),
+        }
+    }
+}
+
+// TODO: When done, remove the aes128gcm prefixes and the EC_ ones.
+// As for now it makes it easier to Ctrl + F into ecec :)
+
+pub struct Aes128GcmEceWebPush<L, R, C> {
```
Comment (eoger) [inline src/aes128gcm.rs:54]: The use of generics and associated types in this struct and the trait it implements is 100% weird, but somehow it's the only way I've found to make it work. Maybe there's a better way of doing that?

File: src/aes128gcm.rs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+            &key,
+            auth_secret,
+            salt,
+            rs,
+            ciphertext,
+        )
+    }
+}
+
+impl<L, R, C> EceWebPush for Aes128GcmEceWebPush<L, R, C>
```
Comment (eoger) [inline src/aes128gcm.rs:156]: Brrrr

**Comment by eoger** [comment]
This is the architecture I came up with, let me know if it actually makes sense:

rust-ece-arch

File: crypto/openssl/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "ece-crypto-openssl"
+version = "0.1.0"
+authors = ["Edouard Oger <eoger@fastmail.com>"]
+license = "MPL-2.0"
+
+[dependencies]
+hkdf = "0.6"
```
Comment (vladikoff) [inline crypto/openssl/Cargo.toml:8]: hkdf 0.7 is live now

File: crypto/openssl/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+lazy_static = "1.1.0"
+openssl = "0.10"
+sha2 = "0.7"
```
Comment (vladikoff) [inline crypto/openssl/Cargo.toml:11]: can use a more recent version now, see comment above

**Comment by eoger** [comment]
Let's re-open this another day

**Pull request closed without merging**